### DISCUSSION TITLE
Error: An admin for model "User" has to be registered to be referenced by TokenAdmin.autocomplete_fields.

### DISCUSSION THREAD

**Discussion by kevinrenskers** [encode/django-rest-framework#9296] [Potential Issue] [answered]
After updating to DRF 3.15.0 I am now getting the following error when starting my Django server:

`<class 'rest_framework.authtoken.admin.TokenAdmin'>: (admin.E039) An admin for model "User" has to be registered to be referenced by TokenAdmin.autocomplete_fields.`

I am using a custom AdminSite instance where only the ModelAdmins that I specifically register to it are shown in my admin site. I have not registered TokenAdmin to my AdminSite, because it's not something we want to see in our admin site. I also have a custom User model, but its ModelAdmin is registered with my custom AdminSite.

It seems that DRF can't handle this situation. I guess that it only looks through the default AdminSite for what is registered there? How can I make DRF just ignore the TokenAdmin? I don't want it in our admin site, as we don't use tokens (only sessions).

**Comment by browniebroke** [comment]
You can unregister the models you don't want with:

```python
from django.contrib import admin
from rest_framework.authtoken.models import TokenProxy


admin.site.unregister(TokenProxy)
```

https://docs.djangoproject.com/en/5.0/ref/contrib/admin/#django.contrib.admin.AdminSite.unregister

You can probably do that anywhere, but it might have to be in an `admin.py` file.

**Reply by i-huzaifa-arshad** [reply]
It worked for me. Idk why but unregistering the `TokenProxy` solved the error. Whereas the error was about `TokenAdmin`. Any insights on this?

**Reply by browniebroke** [reply]
Because DRF registers the model `TokenProxy` to use the model admin `TokenAdmin`:

https://github.com/encode/django-rest-framework/blob/337ba211e82628c0a0bfc7340c45b22e725f8162/rest_framework/authtoken/admin.py#L55

To unregister, the model class is enough, you don't need to specify the admin class.

**Reply by i-huzaifa-arshad** [reply]
My error code is slightly different.
```
<class 'rest_framework.authtoken.admin.TokenAdmin'>: (admin.E040) ModelAdmin must define "search_fields", because it's referenced by TokenAdmin.autocomplete_fields.
```

This is a strange fix in one way. Unregistering `TokenProxy` fixes the issue. But it removes the `Token` page from admin panel. 
```
from rest_framework.authtoken.models import TokenProxy
admin.site.unregister(TokenProxy)
```
![image](https://github.com/encode/django-rest-framework/assets/76521534/7a877782-39fb-4c8c-8b3f-e560ac231116)

Most of my API endpoints require a token for authentication. So, it's not a good fix for my case. For me what worked is the following code (I would rather call it a pseudo-fix):
```
from rest_framework.authtoken.models import TokenProxy
admin.site.unregister(TokenProxy)
admin.site.register(TokenProxy)
```

![image](https://github.com/encode/django-rest-framework/assets/76521534/ffb0b6ae-167f-422a-bd4b-d0643a1f0745)

**Comment by kevinrenskers** [comment]
That doesn't work: `django.contrib.admin.exceptions.NotRegistered: The model TokenProxy is not registered`

I am not using the default admin site, I am not using its auto discover function, so nothing should be registered to it. And yeah, thus this error.

**Reply by browniebroke** [reply, marked as answer]
When you got this error, did you call `unregister` on you custom admin site or on the default admin site? [DRF registers on the default site](https://github.com/encode/django-rest-framework/blob/337ba211e82628c0a0bfc7340c45b22e725f8162/rest_framework/authtoken/admin.py#L55), so even though you're not using it, perhaps it's making the Django checks fail?

> I don't want it in our admin site, as we don't use tokens (only sessions)

Do you have `'rest_framework.authtoken'` in your `INSTALLED_APPS`? If you're not using Tokens, I guess you shouldn't have it, but perhaps it's there?

**Reply by kevinrenskers** [reply]
I called `unregister` on the default admin site.

But I did have `rest_framework.authtoken` in the `INSTALLED_APPS`, yes. I've removed that and the problem is gone.

I do think there's still a problem with DRF and custom admin sites though, right? If you can't remove it from `INSTALLED_APPS`, then I guess you're forced to register your User admin with the default admin site, even though it's not being used?

**Reply by browniebroke** [reply]
Interesting... Are you using a custom user model? [Django also registers the default `User` model](https://github.com/django/django/blob/10b31eea337261ba724a71839ebc4bd4ddac3675/django/contrib/auth/admin.py#L43), so I guess DRF assumes that it's always there, but with the custom User model it might not...

> I do think there's still a problem with DRF and custom admin sites though, right? If you can't remove it from `INSTALLED_APPS`, then I guess you're forced to register your User admin with the default admin site, even though it's not being used?

Yes, ICYMI, there is now an issue #9300

**Reply by kevinrenskers** [reply]
Yes, I also have a custom User model (its ModelAdmin is registered with my custom AdminSite).
